src/multicore2.rs

```diff
@@ -0,0 +1,132 @@
+use std::collections::VecDeque;
+use std::sync::Arc;
+
+use parking_lot::{Condvar, Mutex};
+
+use crossbeam::channel::{self, Receiver};
+
+type Task = Box<dyn FnOnce() + Send + 'static>;
+
+#[derive(Default)]
+struct TaskQueue(VecDeque<Task>, VecDeque<Task>);
+
+#[derive(Default)]
+struct State(Mutex<TaskQueue>, Condvar);
+
+#[derive(Clone)]
+pub struct Worker {
+    pub(crate) cpus: usize,
+    shared_state: Arc<State>,
+}
+
+impl Worker {
+    fn spawn_in_pool(shared_state: Arc<State>) {
+        std::thread::spawn(move || {
+            let State(ref mutex, ref cv) = &*shared_state;
+            let mut guard = mutex.lock();
+
+            loop {
+                while let Some(task) = guard.0.pop_front() {
+                    drop(guard);
+                    task();
+                    guard = mutex.lock();
+                }
+
+                match guard.1.pop_front() {
+                    Some(task) => {
+                        drop(guard);
+                        task();
+                        guard = mutex.lock();
+                    }
+                    None => cv.wait(&mut guard),
+                }
+            }
+        });
+    }
+
+    fn start_workers(cpus: usize, shared_state: &Arc<State>) {
+        for _ in 0..cpus {
+            Self::spawn_in_pool(shared_state.clone());
+        }
+    }
+
+    pub fn new_with_cpus(cpus: usize) -> Worker {
+        assert!(cpus > 0);
+
+        let shared_state = Arc::default();
+        Self::start_workers(cpus, &shared_state);
+        Worker { cpus, shared_state }
+    }
+
+    pub fn new() -> Worker {
+        Self::new_with_cpus(num_cpus::get())
+    }
+
+    pub fn compute<F, T, E>(&self, f: F) -> Receiver<Result<T, E>>
+    where
+        F: FnOnce() -> Result<T, E> + Send + 'static,
```

Comment:
The same way, F doesn't have any access to the resource constraining capabilities. It should get some "spawner", or "scope", or whatever, or at least immediately supply an number of CPU cores it wants to utilize

Reply:
It's supposed to be used for executing subtasks, i.e. first you divide the computation into closures and then pass them to the threadpool.
We can give an access to the scope in the closure, but the problem is in FIFO task processing.
While rayon and crossbeam use local thread queues, and spawning new tasks from scope guarantees that all subtasks are going to be executed next, with our global queue this is not the case. If there're tasks awaiting in the queue, those you're spawn will screw up the order (either postponed to the queue tail or coming to the head in front of scheduled ones)

Reply:
But I totally missed this use-case, my bad, will look into it